source/flux/parts/sis.js

```diff
@@ -39,7 +44,42 @@ export function updateBalances(
 	}
 }
 
-type Action = UpdateBalancesActions
+type LoadCachedCoursesAction = {|
+	type: 'sis/LOAD_CACHED_COURSES',
+	payload: Array<CourseType>,
+|}
+
+export type CourseDataActionType = ThunkAction<CourseDataAction>
+export function getCachedCourses(): CourseDataActionType {
+	return async dispatch => {
+		await updateStoredCourses()
+		const cachedCourses = await loadCachedCourses()
+		dispatch({type: LOAD_CACHED_COURSES, payload: cachedCourses})
+	}
+}
+
+type TermsUpdateStartAction = {|type: 'sis/TERMS_UPDATE_START'|}
+
+type TermsUpdateCompleteAction = {|type: 'sis/TERMS_UPDATE_COMPLETE'|}
+
+export function updateCourseData(): CourseDataActionType {
+	return async dispatch => {
+		dispatch({type: TERMS_UPDATE_START})
+		let updateNeeded = await updateStoredCourses()
+		if (updateNeeded) {
+			const cachedCourses = await loadCachedCourses()
+			dispatch({type: LOAD_CACHED_COURSES, payload: cachedCourses})
```

Comment:
I don't quite follow the logic here – if we need to update the courses, we … return the cached courses?

Reply:
Yeah the call to updateStoredCourses() checks to see if the cached courses are up to date, and updates them if not. It returns true if any of the cached courses were updated, hence the need for the cached courses to be loaded again.

Reply:
ah